### PR TITLE
Jointformer development

### PR DESCRIPTION
**Pull request by AdamIzdebski**

### CODE REVIEW

File: jointformer/models/jointformer.py [hunk 1]
```diff
@@ -59,44 +72,53 @@ def __init__(
...
+        
+        outputs = super().forward(input_ids=input_ids, attention_mask=_attention_mask, is_causal=_is_causal)
+        if _is_causal:
+            if next_token_only:
+                outputs["logits_generation"] = self.lm_head(outputs['embeddings'][:, [-1], :])
+            else:
+                outputs["logits_generation"] = self.lm_head(outputs["embeddings"])
+        else:
+            outputs["logits_physchem"] = self.physchem_head(outputs['embeddings'][:, 0, :])
+            outputs["logits_prediction"] = self.prediction_head(outputs['embeddings'][:, 0, :])
```
Comment (jano1906) [inline jointformer/models/jointformer.py:97]: I don't see that we append any cls token to the input so why do we arbitrarly treat the first token as cls?
Also, ideally we would have 2 separate cls tokens for physchem and prediction

File: jointformer/models/transformer.py [hunk 2]
```diff
@@ -1,74 +1,55 @@
...
+                )
+              for _ in range(self.num_layers)])
+        self.layer_norm = RMSNorm(self.embedding_dim, self.layer_norm_eps)
 
     def forward(
             self,
             input_ids: Optional[torch.Tensor] = None,
             attention_mask: Optional[torch.Tensor] = None,
-            is_causal: Optional[bool] = True
+            is_causal: Optional[bool] = True,
```
Comment (jano1906) [inline jointformer/models/transformer.py:44]: This `is_causal` flag is so bug prone I would set it as a nonoptional keyword argument.
`
def forward(
            self,
            input_ids: torch.Tensor,
            attention_mask: torch.Tensor,
            *
            is_causal: bool,
`

Comment (jano1906) [reply jointformer/models/transformer.py:44]: And if any default value this would have, it should be False since True changes the default behaviour (ignores one of the input)

File: jointformer/models/jointformer.py [hunk 3]
```diff
@@ -59,44 +72,53 @@ def __init__(
     def forward(
             self,
             input_ids: Optional[torch.Tensor] = None,
-            labels: Optional[torch.Tensor] = None,
             attention_mask: Optional[torch.Tensor] = None,
```
Comment (jano1906) [inline jointformer/models/jointformer.py:75]: I would be more calm if attention mask was non optional argument and assert(attention_mask is not None) would be placed at the top of the function. Since we always use this with tokenizer and tokenizer always produces the attention mask it shouldn't be an inconvenient

File: jointformer/models/jointformer.py (hunk 1 above)
Comment (jano1906) [reply jointformer/models/jointformer.py:97]: Also if we had the cls token we wouldn't need to aggregate the tokens used for language modeling as it is done now.

File: jointformer/models/jointformer.py [hunk 4]
```diff
@@ -105,61 +127,56 @@ def get_loss(
...
+            attention_mask: Optional[torch.Tensor] = None,
+            input_labels: Optional[torch.Tensor] = None,
+            **kwargs
+            ):
 
+        outputs = self.forward(input_ids=input_ids, attention_mask=attention_mask, task='mlm')
+        outputs["logits_generation"] = self.mlm_head(outputs['embeddings'])
+        if input_labels is not None:
+            outputs["loss"] = mlm_loss(outputs["logits_generation"], input_labels, IGNORE_INDEX, 'mean')
         return outputs
```
Comment (jano1906) [inline jointformer/models/jointformer.py:152]: btw, when do we do masked language modeling? This also would require special mask token for training.

File: jointformer/models/jointformer.py [hunk 5]
```diff
@@ -105,61 +127,56 @@ def get_loss(
...
-
+            if self.prediction_task_type == 'classification':
+                outputs["loss"] = classification_loss(outputs["logits_prediction"], properties, reduction='mean')
+            elif self.prediction_task_type == 'regression':
+                outputs["loss"] = regression_loss(outputs["logits_prediction"], properties, 'mean')
+            else:
+                raise ValueError('Variable `prediction_task_type` must be either `classification` or `regression`.')
         return outputs
 
     def calculate_perplexity(
```
Comment (jano1906) [inline jointformer/models/jointformer.py:182]: Remember this should be only calculated in eval mode. Btw, some discussion: https://discuss.pytorch.org/t/opinion-eval-should-be-a-context-manager/18998

File: jointformer/configs/dataset.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        self.validate = validate
+        self.standardize = standardize
+        self.num_samples = num_samples
+        self.split = split
+        self.splitter = splitter
+        self.transform = transform
+        self.target_transform = target_transform
+        self.task_type = task_type
+        self.metric = metric
+        self.num_tasks = num_tasks
```
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/configs/dataset.py:51]: I leave it for future consideration, but in general yes

[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/configs/dataset.py:51]: #28

File: jointformer/models/base.py [hunk 7]
```diff
@@ -12,6 +16,33 @@ def to_guacamole_generator(self, tokenizer, batch_size, temperature, top_k, devi
...
+    
+    @abc.abstractmethod
+    def predict(
+            self,
+            input_ids: Optional[torch.Tensor] = None,
+            attention_mask: Optional[torch.Tensor] = None,
+            **kwargs
+           ):
+        pass
+
```
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/base.py]: added

File: jointformer/models/layers/attention.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        self.dropout = dropout
+        self.block_size = block_size
+        
+        # key, query, value projections for all heads, but in a batch
+        self.qkv = nn.Linear(self.embedding_dim, 3 * self.embedding_dim, bias=bias)
+        self.out = nn.Linear(self.embedding_dim, self.embedding_dim, bias=bias)
+        self.relative_embedding = RotaryPositionalEmbedding(self.head_dim)
+  
+        if not hasattr(torch.nn.functional, 'scaled_dot_product_attention'):
+            raise ImportError("The `scaled_dot_product_attention` function is not available in PyTorch < 2.0.0.")
```
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/layers/attention.py:29]: changed

File: jointformer/models/layers/layer_norm.py [hunk 9]
```diff
@@ -3,30 +3,16 @@
...
-    def __init__(self, ndim: int, eps: float):
+    def __init__(self, embedding_dim: int, epsilon: float = 1e-5):
         super().__init__()
-        self.weight = nn.Parameter(torch.ones(ndim))
-        self.eps = eps
+        self.epsilon = epsilon
+        self.weight = nn.Parameter(torch.ones(embedding_dim))
 
     def _norm(self, x):
-        return x * torch.rsqrt(torch.mean(x * x, dim=-1, keepdim=True) + self.eps)
```
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/layers/layer_norm.py:28]: changed

File: jointformer/models/layers/prediction.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class RegressionHead(nn.Module):
+
+    def __init__(self, embedding_dim: int, prediction_hidden_dim: int, output_dim: int):
+        super().__init__()
+        self.net = nn.Sequential(
+            nn.Linear(embedding_dim, prediction_hidden_dim),
+            nn.GELU(),
+            nn.LayerNorm(prediction_hidden_dim),
```
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/layers/prediction.py:12]: Actually, it probably should. 

Reference, for the current version: https://github.com/evolutionaryscale/esm/blob/main/esm/layers/regression_head.py

We can discuss, or benchmark

File: jointformer/models/utils.py [hunk 11]
```diff
@@ -29,3 +33,55 @@ def generate(self, number_samples: int) -> List[str]:
...
+            attention_mask=attention_mask
+            )
+     
+    @property
+    def global_embedding(self):
+        if self.attention_mask is not None and self.embedding is not None:
+            return torch.einsum('bsd, bs -> bsd', self.embedding, self.mask.float())
+        else:
+            return None
+
```
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/utils.py:71]: changed to follow the pr

File: jointformer/models/utils.py [hunk 12]
```diff
@@ -29,3 +33,55 @@ def generate(self, number_samples: int) -> List[str]:
...
+    labels = labels[:, 1:].contiguous()
+    return F.cross_entropy(logits.view(-1, logits.size(-1)), labels.view(-1), ignore_index=ignore_index, reduction=reduction)
+
+
+def mlm_loss(logits, labels, ignore_index, reduction):
+    return F.cross_entropy(logits.view(-1, logits.size(-1)), labels.view(-1), ignore_index=ignore_index, reduction=reduction)
+
+
+def regression_loss(logits, labels, reduction):
+    return F.mse_loss(logits.flatten(), labels.flatten(), reduction=reduction)
```
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/utils.py:84]: changed, will become an instance of metrics as in #27

File: jointformer/models/jointformer.py [hunk 13]
```diff
@@ -21,35 +27,42 @@ def __init__(
...
-            vocab_size=vocab_size, max_seq_len=max_seq_len, embedding_dim=embedding_dim,
-            dropout=dropout, num_layers=num_layers, bias=bias, num_heads=num_heads, layer_norm_eps=layer_norm_eps)
-
+            vocab_size=vocab_size, max_seq_len=max_seq_len, embedding_dim=embedding_dim, embedding_hidden_dim=embedding_hidden_dim, attention_dropout=attention_dropout,
+            feed_forward_dropout=feed_forward_dropout, num_layers=num_layers, bias=bias, num_heads=num_heads, layer_norm_eps=layer_norm_eps
+            )
         # Hardcoding all tasks into the model definition for easier serialization
+        self.prediction_task_type = prediction_task_type
         self.lm_head = nn.Linear(self.embedding_dim, self.vocab_size, bias=False)
         self.mlm_head = nn.Linear(self.embedding_dim, self.vocab_size, bias=False)
```
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:52]: For some reason it is a convention to do it this way:
https://github.com/karpathy/nanoGPT/blob/9755682b981a45507f6eb9b11eadef8cb83cebd5/model.py#L133
https://github.com/pytorch-labs/gpt-fast/blob/091515ab5b06f91c0d6a3b92f9c27463f738cc9b/model.py#L99

File: jointformer/models/jointformer.py [hunk 14]
```diff
@@ -21,35 +27,42 @@ def __init__(
...
-            nn.Linear(self.embedding_dim, num_physchem_tasks),
-        )
-        self.prediction_head = nn.Linear(self.embedding_dim * max_seq_len, num_prediction_tasks, bias=False)
+        self.physchem_head = RegressionHead(embedding_dim=self.embedding_dim, prediction_hidden_dim=prediction_hidden_dim, output_dim=num_physchem_tasks)
 
+        # Init prediction head depending on task type
+        if prediction_task_type == 'classification':
+            self.prediction_head = nn.Linear(self.embedding_dim, num_prediction_tasks)
+        elif prediction_task_type == 'regression':
+            self.prediction_head = RegressionHead(embedding_dim=self.embedding_dim, prediction_hidden_dim=prediction_hidden_dim, output_dim=num_prediction_tasks)
```
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:59]: Added ClassificationHead

File: jointformer/models/jointformer.py [hunk 15]
```diff
@@ -59,44 +72,53 @@ def __init__(
     def forward(
             self,
             input_ids: Optional[torch.Tensor] = None,
-            labels: Optional[torch.Tensor] = None,
             attention_mask: Optional[torch.Tensor] = None,
-            is_causal: bool = True,
-            next_token_only: bool = False,
+            task: Optional[str] = None,
+            next_token_only: Optional[bool] = False,
```
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:77]: made it required

File: jointformer/models/jointformer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:75]: corrected

File: jointformer/models/jointformer.py (hunk 1 above)
Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:97]: You can have two separate cls tokens by setting set_separate_task_tokens hparam in
https://github.com/szczurek-lab/jointformer/blob/e1ca7f0311da70b6756ba75b51e4f78db3b7794c/configs/tokenizers/smiles/config.json#L5

Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:97]: we can still use the first token as the prediction?

Maybe we should adress this in a separate PR?

File: jointformer/models/transformer.py (hunk 2 above)
Comment (AdamIzdebski) [reply jointformer/models/transformer.py:44]: yes, very True

Comment (AdamIzdebski) [reply jointformer/models/transformer.py:44]: I made it nonoptional

File: jointformer/models/jointformer.py (hunk 3 above)
Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:75]: So, it is actaully None when is_causal = True, no?

File: jointformer/models/jointformer.py (hunk 4 above)
Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:152]: We will train a simply MLM for ablation I guess -> it was used for the previous version of Jointformer

Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:152]: added a separate task token for MLM

File: jointformer/models/jointformer.py (hunk 5 above)
Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:182]: Interesting!

Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:182]: #27 moves calculating perplexity

File: jointformer/models/jointformer.py (hunk 1 above)
Comment (jano1906) [reply jointformer/models/jointformer.py:98]: todo

File: jointformer/models/transformer.py (hunk 2 above)
Comment (jano1906) [reply jointformer/models/transformer.py]: todo

File: jointformer/models/jointformer.py (hunk 3 above)
Comment (jano1906) [reply jointformer/models/jointformer.py:76]: todo

File: jointformer/models/jointformer.py (hunk 1 above)
Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:98]: todo: offset generation forward pass

File: jointformer/models/layers/prediction.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/layers/prediction.py:12]: left unchanged, done

File: jointformer/models/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (AdamIzdebski) [reply jointformer/models/utils.py]: done

File: jointformer/models/jointformer.py (hunk 1 above)
Comment (AdamIzdebski) [reply jointformer/models/jointformer.py:97]: done

File: jointformer/models/transformer.py (hunk 2 above)
Comment (AdamIzdebski) [reply jointformer/models/transformer.py]: done

File: jointformer/models/jointformer.py (hunk 3 above)
Comment (AdamIzdebski) [reply jointformer/models/jointformer.py]: done

**Review by jano1906 — APPROVED** [review]
ok
